packages/flutter_test/lib/src/widget_tester.dart

```diff
@@ -174,9 +174,10 @@ Future<void> expectLater(dynamic actual, dynamic matcher, {
   String reason,
   dynamic skip, // true or a String
 }) {
-  return TestAsyncUtils.guard(() async {
-    await test_package.expectLater(actual, matcher, reason: reason, skip: skip);
-  });
+  // We can't wrap the delegate in a guard, or we'll hit async barriers in
+  // [TestWidgetsFlutterBinding] while we're waiting for the matcher to complete
+  TestAsyncUtils.guardSync();
+  return test_package.expectLater(actual, matcher, reason: reason, skip: skip);
```

Comment:
What I'm worried about is that this will mean you can forget to await this method, then call something like `pump`, and we won't catch it.

Why are we hitting async barriers if you await this future?

Reply:
While still awaiting the async matcher, I was hitting the async barrier here:

https://github.com/flutter/flutter/blob/5972fdb7aa251eeed8cb5dcb74856f47f9570225/packages/flutter_test/lib/src/binding.dart#L485